docarray/utils/apply.py

```diff
@@ -0,0 +1,299 @@
+from contextlib import nullcontext
+from math import ceil
+from multiprocessing.pool import Pool, ThreadPool
+from types import LambdaType
+from typing import Any, Callable, Generator, Optional, TypeVar, Union
+
+from rich.progress import track
+
+from docarray import BaseDocument
+from docarray.array.abstract_array import AnyDocumentArray
+
+T = TypeVar('T', bound=AnyDocumentArray)
+T_doc = TypeVar('T_doc', bound=BaseDocument)
+
+
+def apply(
+    da: T,
+    func: Callable[[T_doc], T_doc],
+    backend: str = 'thread',
```

Comment:
Not sure about the default option. The GIL means that threading is not a good option for many tasks; so mp seems like the safer default choice? Even though there is more overhead?